### ISSUE TITLE
feature request: Exec backup type

### ISSUE THREAD

**Issue by higuita** [backuppc/backuppc#86]
I love backuppc, is very flexible and works very well... but i miss one feature: Exec backup type

I want to execute a remote script to do a backup. The backup will not be store in backuppc, so all i need is to execute a script and save the status and the script log. My most common usage is backup some data offsite to AWS S3

Right now i workaround with tar, i replace the tar command with a script that will execute my command and then tars the log, so backuppc can see the backup status and a user can checkout the log.
It works, but is a pain to be forced to play with tar command, the full and incremental options to do what i need and then fake a tar just to make backuppc happy

A simple ssh $host $script $options would be perfect, where the $options get their parameters from the full and incremental backup. Restore could also be a script. The script status code and stdout and stderr should be captured and logged by backuppc. No files are transfered to backuppc, so the size for this type of backups could always be zero in the status (or show it inside a [], to flag a remote backup)

With this, one could use backuppc as the only interface for all backups. This is useful specially on remote machines with very slow network, or cloud. This could also open the door to use tapes, dvd burners and interact with other backup software

**Comment by ngharo** [comment]
It sounds like you should be using $Conf{DumpPreUserCmd} to execute your command.  Then use BackupPC as intended to backup the results of your command.

**Comment by higuita** [comment]
That is the same "problem"! The command will run the backup but then you need to hack around as you have nothing to backup to backuppc. 

yes, we can workaround by backup /etc/ or some almost empty dir... but this is always hacking.

what i'm requesting is to improve backuppc and add official support for this, everything needed is there, it is mostly copy&paste and strip the extra code for the backup, should not be that hard (says the guy that have only basic programming skills! :D )

**Comment by ngharo** [comment]
What I'm suggesting is your script produce artifacts for BackupPC to backup.

For instance, you mentioned you have something that writes to AWS S3.  Have that script write a log file:

```
#!/bin/bash
# This script is /usr/local/bin/prebackup_command
log_dir=/usr/local/var/mylogs
log_file="${log_dir}/$(date +%y%m%d)-results.log"

echo "Log file opened at $(date)" >> $log_file
something_aws_s3_command >> $log_file 2>&1
echo "Log file closed at $(date)" >> $log_file
```

Now you configure BackupPC (without "hacks") to backup `/usr/local/var/mylogs` with a `DumpPreUserCmd` of `$sshPath -q -x -l root $host /usr/local/bin/prebackup_command`

**Comment by higuita** [comment]
Yes, i understand that... but without all the stdout and stderr redirection, the logs are already in the backuppc log file, so the backup part is useless

Notice that i already have this working, i used tar originally, right now i do use a DumpPreUserCmd with a rsync backup and i backup the /home/backuppc user dir (that only have the .ssh)

Just like we have tar, rsync, rsyncd , ftp, smb backups, i'm suggesting adding a plain "exec"  backup type. This will help people starting with backuppc and everyone in long run.

**Comment by ngharo** [comment]
I see what you mean but your use case has BackupPC acting as a web interface to schedule a command.  Like a cron with a web interface (GUI cron projects already exist). All the existing XferMethods you list all end up producing data.  Your use case does not so I really don't think it fits in.

Please note that I'm just another user like you with no real say into the direction development goes other than expressing my opinions.

**Comment by higuita** [comment]
True, this could be replaced by a cron... but crontab do not have status report, retry, blackout, etc ... or at least not without lot of work...

Having several tools to manage backups is also a pain and promotes errors. Having a central place to setup and monitor backups is a lot better... after all, almost all the backuppc can be replaced by a bunch of scripts running on cron... but backuppc is a LOT better than that

This feature is not useful for everyone, but also smb or ftp is not useful for everyone, but it is there and many people use then

**Comment by SpareSimian** [comment]
ngharo's point was not to use traditional cron (although that could work) but a GUI scheduler. It appears that you're using BackupPC only as a scheduler so you're likely to find something better when it's not devoted to BackupPC's backup model.

**Comment by higuita** [comment]
yes, i understand, but it still another tool to setup, with it own configs, databases and still requires one to monitor backups using different tools. Backuppc can do all this, just being forced to run a backup in the end is not friendly and makes hard for newbies in backuppc to setup those backups

**Comment by ngharo** [comment]
But they're not backups.  

Although.. I think it's really cool and shows how flexible BackupPC is.

 Have you tried poking at the source a bit?  It sounded like you have it already working but want it to be more user friendly. Take a look at `bin/BackupPC_dump`.  It's pretty easy to read.  That will give you a good idea what is happening when backuppc kicks off a backup.  I found it easier to work myself up from there when making a couple specific changes for one of my weird cases.

**Comment by jkroepke** [comment]
We use BackupPC for S3 backup via `DumpPreUserCmd`, too.

The amount of S3 data raises up to multiple GBs. Using `DumpPreUserCmd` then rsync results into double storage size requirements.

There are any idea to archive this without double storage requirements? Test s3fs/fuse but it wont work well...

**Comment by higuita** [comment]
As i said above, you can workaround that by requesting backuppc to backup the /etc/ or a (almost) empty folder (i use sometimes the folder where i place the S3 transfer log) and backuppc will transfer a minimal amount of data. Do not forget to force the DumpPreUserCmd  with the correct list of paths to backup as the backuppc one will now be a wrong folder/file.

I tried to add a new option, but my perl fu isnt' good enough :)

**Comment by jkroepke** [comment]
I got it, but with exec backup type, I understood to backup a S3 Bucket inside backuppc using aws s3 command. 

Use backup to just run an external script... is an other use case...

**Comment by GWHAYWOOD** [comment]
Given the available development effort it seems unlikely that something like this will be implemented in the foreseeable future.

**Comment by GWHAYWOOD** [comment]
Closed as not achievable.
